integration/ec2_test.go

```diff
@@ -139,20 +170,101 @@ func TestSimplifiedNodeJoin(t *testing.T) {
 	require.NoError(t, err)
 	require.Empty(t, nodes)
 
 	// create and start the node
-	nodeConfig := newNodeConfig(t, authConfig.Auth.SSHAddr, tokenName)
+	nodeConfig := newNodeConfig(t, authConfig.Auth.SSHAddr, tokenName, types.JoinMethodEC2)
 	nodeSvc, err := service.NewTeleport(nodeConfig)
 	require.NoError(t, err)
 	require.NoError(t, nodeSvc.Start())
 	t.Cleanup(func() { require.NoError(t, nodeSvc.Close()) })
 
 	// the node should eventually join the cluster and heartbeat
 	require.Eventually(t, func() bool {
-		nodes, err := authServer.GetNodes(context.Background(), defaults.Namespace)
+		nodes, err := authServer.GetNodes(context.Background(), apidefaults.Namespace)
+		require.NoError(t, err)
+		return len(nodes) > 0
+	}, time.Minute, time.Second, "waiting for node to join cluster")
+}
+
+// TestIAMNodeJoin is an integration test which asserts that the IAM method for
+// Simplified Node Joining works when run on a real EC2 instance with an
+// attached IAM role.  This is a very basic test, unit testing with mocked AWS
+// endpoints is in lib/auth/join_iam_test.go
+func TestIAMNodeJoin(t *testing.T) {
+	if os.Getenv("TELEPORT_TEST_EC2") == "" {
+		t.Skipf("Skipping TestIAMNodeJoin because TELEPORT_TEST_EC2 is not set")
+	}
+
+	// create and start the auth server
+	authConfig := newAuthConfig(t, nil)
+	authSvc, err := service.NewTeleport(authConfig)
+	require.NoError(t, err)
+	require.NoError(t, authSvc.Start())
+
+	authServer := authSvc.GetAuthServer()
+
+	// fetch the caller identity to find the AWS account and create the token
+	id := getCallerIdentity(t)
+
+	tokenName := "test_token"
+	token, err := types.NewProvisionTokenFromSpec(
+		tokenName,
+		time.Now().Add(time.Hour),
+		types.ProvisionTokenSpecV2{
+			Roles: []types.SystemRole{types.RoleNode, types.RoleProxy},
+			Allow: []*types.TokenRule{
+				&types.TokenRule{
+					AWSAccount: *id.Account,
+				},
+			},
+			JoinMethod: types.JoinMethodIAM,
+		})
+	require.NoError(t, err)
+
+	err = authServer.UpsertToken(context.Background(), token)
+	require.NoError(t, err)
+
+	// sanity check there are no proxies to start with
+	proxies, err := authServer.GetProxies()
+	require.NoError(t, err)
+	require.Empty(t, proxies)
+
+	// create and start the proxy, will use the IAM method to join by connecting
+	// directly to the auth server
+	proxyConfig := newProxyConfig(t, authConfig.Auth.SSHAddr, tokenName, types.JoinMethodIAM)
+	proxySvc, err := service.NewTeleport(proxyConfig)
+	require.NoError(t, err)
+	require.NoError(t, proxySvc.Start())
+
+	// the proxy should eventually join the cluster and heartbeat
+	require.Eventually(t, func() bool {
+		proxies, err := authServer.GetProxies()
+		require.NoError(t, err)
+		return len(proxies) > 0
+	}, time.Minute, time.Second, "waiting for proxy to join cluster")
+
+	// InsecureDevMode needed for node to trust proxy
+	lib.SetInsecureDevMode(true)
+	defer lib.SetInsecureDevMode(false)
```

Comment:
Should we reset to whatever the original value was, instead of assuming it was false?